### PR TITLE
2pc: reduce memory footprint by using a more compact mutations list

### PR DESCRIPTION
**Pull request by bobotu**
### What problem does this PR solve?
https://github.com/pingcap/tidb/pull/19055 has too many changes, this PR propose a suboptimal solution to reduce the memory footprint.

### What is changed and how it works?
The rationale behind #19055 is that `CommitterMutations` uses a lot of memory and that is unnecessary. So #19055 try to rewrite the whole 2pc executor to remove the `CommitterMutations` entirely. 

For a suboptimal solution, we can reduce the memory footprint of `CommitterMutaions`. For each mutation in the old `CommitterMutaions` it will use `4 + 1 + 24 + 24 = 53` byte. 

In fact, we do not need to hold all `key` and `value` slice in an array, we just need to save a pointer that can quickly access the data and load them when necessary. When can use the logical pointer of the red-black tree's node to reference `key` and `value`, and due to the fact that there is 16-bit unused space in this pointer, we can encode `op` and `isPessimisticLock` into here. We only need 8 bytes for each mutation.

#### This PR
图片

#### #19055
图片

In this case, the actual size of mutations list is around 830Mb.

### Release note 

- Reduce the memory footprint of the 2PC committer.

### CODE REVIEW

**Comment by bobotu** [comment]
We can close #20370 if we perfer this solution.

**Comment by coocood** [comment]
This solution is much better.

File: store/tikv/2pc.go [hunk 1]
```diff
@@ -116,26 +116,95 @@ type twoPhaseCommitter struct {
...
 	ops               []pb.Op
 	keys              [][]byte
 	values            [][]byte
 	isPessimisticLock []bool
 }
 
-// NewCommiterMutations creates a CommitterMutations object with sizeHint reserved.
-func NewCommiterMutations(sizeHint int) CommitterMutations {
-	return CommitterMutations{
+// NewPlainMutations creates a CommitterMutations object with sizeHint reserved.
```
Comment (coocood) [inline store/tikv/2pc.go:195]: Need update comment

File: kv/memdb.go [hunk 2]
```diff
@@ -125,6 +125,17 @@ var tombstone = []byte{}
 // IsTombstone returns whether the value is a tombstone.
 func IsTombstone(val []byte) bool { return len(val) == 0 }
 
+// MemKeyHandle represents a pointer for key in MemBuffer.
+type MemKeyHandle struct {
+	Passthrough uint16
```
Comment (coocood) [inline kv/memdb.go:130]: How about `Meta`? 
I think is more commonly used and easier to understand.

Comment (bobotu) [reply kv/memdb.go:130]: It is an opaque user data instead of metadata of handle. Similar concepts in kqueue and epoll are referred to as udata or data.

**Comment by coocood** [comment]
LGTM

**Comment by tiancaiamao** [comment]
LGTM

**Comment by bobotu** [comment]
PTAL @lysu @jackysp

**Review by jackysp — APPROVED** [review]
LGTM

**Pull request merged**